aten/src/ATen/native/cpu/Activation.cpp

```diff
@@ -589,6 +596,111 @@ void glu_backward_kernel(TensorIterator& iter) {
   });
 }
 
+#if AT_MKL_ENABLED()
+
+template <typename T>
+void SiLUMKLKernel(TensorIterator* it) {
+  if (!it->can_use_32bit_indexing()) {
+    for (auto& sub_it : it->with_32bit_indexing()) {
+      SiLUMKLKernel<T>(&sub_it);
+    }
+    return;
+  }
+
+  constexpr int64_t K = Vec256<T>::size();
+  const int64_t N = it->numel();
+  const T* X_data = static_cast<T*>(it->data_ptr(2));
+  T* Y_data = static_cast<T*>(it->data_ptr(0));
+  T* buffer_data = static_cast<T*>(it->data_ptr(1));
+  at::parallel_for(0, N, K, [=](int64_t begin, int64_t end) {
+    for (int64_t i = begin; i < end; ++i) {
+      buffer_data[i] = -X_data[i];
+    }
+    MKLExp<T>(end - begin, buffer_data + begin, buffer_data + begin);
+    for (int64_t i = begin; i < end; ++i) {
+      const T sigmoid = T(1) / (T(1) + buffer_data[i]);
+      Y_data[i] = X_data[i] * sigmoid;
+      buffer_data[i] = sigmoid;
+    }
+  });
+}
+
+#else
+
+template <typename T>
+void SiLUMKLKernel(TensorIterator* it) {
+  AT_ASSERTM(false, "ATen not compiled with MKL");
+}
+
+#endif // AT_MKL_ENABLED
+
+void silu_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(
+      kBFloat16, iter.dtype(), "silu_cpu", [&]() {
+        if (at::hasMKL() && iter.is_contiguous()) {
+          SiLUMKLKernel<scalar_t>(&iter);
+        } else {
+          Tensor input = iter.input(0);
+          Tensor output = iter.output(0);
+          Tensor buffer = iter.output(1);
+          auto iter1 = TensorIterator::unary_op(buffer, input);
+          auto iter2 = TensorIterator::binary_op(output, input, buffer);
+          const Vec256<scalar_t> kOneVec(scalar_t(1));
+          cpu_kernel_vec(
+              iter1,
+              [](scalar_t x) {
+                return scalar_t(1) / (scalar_t(1) + std::exp(-x));
+              },
+              [kOneVec](Vec256<scalar_t> x_vec) {
+                return kOneVec / (kOneVec + x_vec.neg().exp());
+              });
+          cpu_kernel_vec(
+              iter2,
+              [](scalar_t a, scalar_t b) { return a * b; },
+              [](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a * b; });
+        }
+      });
+}
+
+template <typename T>
+void SiLUBackwardContiguousKernel(TensorIterator* it) {
```

Comment:
why do some kernels take TensorIterators as references and some as points?

Reply:
Removed.